### PR TITLE
Homework2.KNN.Kushnir

### PR DESCRIPTION
**Pull request by ArtemKushnir**
notebook: https://colab.research.google.com/drive/1tYM911_0aNZLj7ylCuDvc2OoD3NQJ99B?usp=sharing
Выводы: 
**spam.csv** 
spam_result

**Лучшие метрики на валидационной выборке:** 

- accuracy=0.91, f1-score - 0.88 (metric=manhattan, k=15)

**Результаты на тестовой выборке:**

-  accuracy=0.81, f1-score=0.75

Результаты валидации указывают на наличие выбросов в данных, поскольку манхэттенская метрика показала себя лучше. Это также подтверждается сильными колебаниями метрик при малых k.

Можно предположить, что объекты одного класса могут существенно отличаться по небольшому числу признаков. Это объясняет, почему евклидова метрика показала лучшие результаты при больших k (41–57), тогда как манхэттенская метрика меньше штрафовала за сильные отклонения по отдельным признакам.

На графике видно, что метрики на валидационной выборке растут при k от 1 до 15, а затем постепенно снижаются. Это говорит о том, что соседние объекты действительно часто принадлежат к одному классу, но эта зависимость выражена не очень сильно, поскольку метрики уменьшаются плавно, а не резко.

**cancer.csv**
cancer_result

**Лучшие метрики на валидационной выборке:** 

- accuracy=0.99, f1-score - 0.99 (metric=manhattan, k=7)

**Результаты на тестовой выборке:**

-  accuracy=0.97, f1-score=0.96

График показывает, что объекты двух классов хорошо разделимы в пространстве. Даже при больших k knn очень хорошо распознает объекты. 

Высокая точность на тестовой выборке указывает на то, что модель хорошо обобщает данные.

Метрики остаются стабильно высокими при любом k, поэтому данные возможно линейно разделимы.

### CODE REVIEW

File: src/homework2/kd_tree.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+class MaxHeap:
+    def __init__(self, k: int) -> None:
+        self.heap: list[Neighbour] = []
+        self.size: int = 0
+        self.k: int = k
+
+    def push(self, neighbour: Neighbour) -> None:
+        neighbour.distance *= -1
+        if len(self.heap) < self.k:
```
Comment (maybenotilya) [inline src/homework2/kd_tree.py]:
```suggestion
        if self.size < self.k:
```

File: src/homework2/kd_tree.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        self.size -= 1
+        neighbour.distance *= -1
+        return neighbour
+
+    def get_k_min(self) -> list[Neighbour]:
+        k_min_neighbour = [self.pop() for _ in range(self.k)][::-1]
+        self.size = 0
+        return k_min_neighbour
+
+    def get_max_element(self) -> Optional[float]:
```
Comment (maybenotilya) [inline src/homework2/kd_tree.py:75]: Судя по названию оно должно возвращать не расстояние, а элемент

File: src/homework2/kd_tree.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+
+class KDTree:
+    _METRICS = {
+        "euclidean": lambda a, b: float(np.linalg.norm(a - b)),
+        "manhattan": lambda a, b: np.sum(np.abs(a - b)),
+        "cosine": lambda a, b: 1 - np.dot(a, b) / (np.linalg.norm(a) * np.linalg.norm(b)),
+    }
+
+    def __init__(self, X: list[Point], leaf_size: int = 1, metric: str = "euclidean") -> None:
```
Comment (maybenotilya) [inline src/homework2/kd_tree.py:86]: Хочется конечно уметь передавать произвольную метрику, но да ладно

File: src/homework2/processing.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    @abstractmethod
+    def transform(self, data: NDArray[np.float64]) -> NDArray[np.float64]:
+        pass
+
+    def fit_transform(self, data: NDArray[np.float64]) -> NDArray[np.float64]:
+        self.fit(data)
+        return self.transform(data)
+
+    def _is_fit(self) -> None:
```
Comment (maybenotilya) [inline src/homework2/processing.py:65]: Если уж решил сделать так, то в идеале сделать `fit` таким, чтобы он вызывал какой нибудь `fit_internal` и делал `fitted = True`, чтобы не забывать это делать в переопределениях `fit`

File: src/homework2/processing.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        return X[indexes[:train_cnt]], X[indexes[train_cnt:]], y[indexes[:train_cnt]], y[indexes[train_cnt:]]
+
+    train_cnt_classes = {key: max(1, int(value * train_cnt / objects_cnt)) for key, value in Counter(y).items()}
+    curr_cnt_classes: defaultdict[T, int] = defaultdict(int)
+
+    train_mask = np.zeros(objects_cnt, dtype=bool)
+    for i in range(objects_cnt):
+        if curr_cnt_classes[y[i]] < train_cnt_classes[y[i]]:
+            train_mask[i] = True
+            curr_cnt_classes[y[i]] += 1
```
Comment (maybenotilya) [inline src/homework2/processing.py:143]: Охх, я никогда не понимал как работает stratify поэтому поверю тебе

File: tests/hoomeworks/homework2/test_proccessing.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+
+@pytest.mark.parametrize("scaler_class", [StandardScaler, MinMaxScaler, MaxAbsScaler])
+def test_scalers(scaler_class):
+    data = np.random.rand(100, 5) * 100
+    scaler = scaler_class()
+    transformed = scaler.fit_transform(data)
+    assert transformed.shape == data.shape
+
+    if isinstance(scaler, StandardScaler):
```
Comment (maybenotilya) [inline tests/hoomeworks/homework2/test_proccessing.py:29]: Ну вообще можно было и в отдельных тестах это сделать

File: src/homework2/processing.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    def _is_fit(self) -> None:
+        if not hasattr(self, "fitted") or not self.fitted:
+            raise ValueError("Scaler is not fitted. Call 'fit()' first or call 'fit_transform()'")
+
+
+class StandardScaler(Scaler):
+    def __init__(self) -> None:
+        self.std: Optional[NDArray[np.float64]] = None
+        self.mean: Optional[NDArray[np.float64]] = None
+        self.fitted: bool = False
```
Comment (maybenotilya) [inline src/homework2/processing.py:74]: Можно вынести в `Scaler` и делать `super().__init__()`

**Review by maybenotilya — COMMENTED** [review]
Как бы я не старался докопаться, тут просто не к чему. Очень хорошо, молодец

**Comment by maybenotilya** [comment]
Ток диффы в ПРе бы делать получше, чтобы не появлялось всяких удалений `hello_world.py`
